### ISSUE TITLE
Union initialised with common value evaluates to first type

### ISSUE THREAD

**Issue by ooliver1** [microsoft/pyright#4660]
**Describe the bug**
An assignment to a union of (the same) generic types causes an interence to the first type.

**To Reproduce**

1. Use a union of common types
2. Assign it to a value that can be interpreted as either
3. Try to use it like it is one of those types (generic method, comparison)

**Expected behavior**
`lst` would be believed to be either `list[str]` or `list[int]` until type checked otherwise (member check for example)

**Screenshots or Code**
```py
lst: list[str] | list[int] = []
lst.append(1)
```

**VS Code extension or command-line**
Both Pylance and Pyright. Pyright 1.1.194, Pylance 2023.2.31.

**Additional context**
This also occurs with `typing.List` + `typing.Union`, even back in 1.1.85 but not 1.1.75.

Here are the results of a `git bisect` between these revisions:

```
01c39df123c8eb5d1042edba0d431d6e32ce096f is the first bad commit
commit 01c39df123c8eb5d1042edba0d431d6e32ce096f
Author: Eric Traut <erictr@microsoft.com>
Date:   Wed Sep 30 10:16:58 2020 -0700

    Improved handling of bidirectional type inference when RHS of assignment is a constructor. More changes to come on this front.

 .../pyright-internal/src/analyzer/typeEvaluator.ts |  74 +++++++-----
 .../pyright-internal/src/tests/checker.test.ts     |   4 +-
 .../src/tests/samples/constructor2.py              | 128 +++++++++++++++++++++
 .../src/tests/samples/genericTypes14.py            |  12 +-
 .../src/tests/samples/typedDict6.py                |   5 +-
 .../src/tests/samples/unnecessaryCast1.py          |  11 +-
 6 files changed, 190 insertions(+), 44 deletions(-)
 create mode 100644 packages/pyright-internal/src/tests/samples/constructor2.py
 ```

**Comment by erictraut** [comment]
The current behavior is "as designed", so I don't consider this a bug. The type of `lst` is narrowed upon assignment. The narrowing logic uses bidirectional type inference to determine the type. When the "expected type" is a union, the bidirectional type inference logic determines which subtypes are potentially compatible. If more than one subtype is potentially compatible, it uses the first one for inference. In your case, `list[int]` and `list[float]` are both potentially compatible, so it uses `list[int]` for bidirectional type inference. The resulting narrowed type after assignment is `list[int]`.

Mypy does something similar in this case.

Even if `lst` were not narrowed to `list[int]` and remained `list[int] | list[str]` after the assignment, the expression `lst.append(1)` would still result in a type violation error because `1` is valid only for one of the two subtypes in the union.

**Comment by erictraut** [comment]
For additional details, you may find the following documentation useful:
[Type checking concepts](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md)
[Type inference](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md)
